### PR TITLE
CA-948 project per workspace db columns and plumbing

### PR DESCRIPTION
**Pull request by dvoet**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 1]
```diff
@@ -386,11 +390,11 @@ trait WorkspaceComponent {
...
     }
 
     private def unmarshalWorkspace(workspaceRec: WorkspaceRecord, attributes: AttributeMap): Workspace = {
-      Workspace(workspaceRec.namespace, workspaceRec.name, workspaceRec.id.toString, workspaceRec.bucketName, workspaceRec.workflowCollection, new DateTime(workspaceRec.createdDate), new DateTime(workspaceRec.lastModified), workspaceRec.createdBy, attributes, workspaceRec.isLocked)
+      Workspace(workspaceRec.namespace, workspaceRec.name, workspaceRec.id.toString, workspaceRec.bucketName, workspaceRec.workflowCollection, new DateTime(workspaceRec.createdDate), new DateTime(workspaceRec.lastModified), workspaceRec.createdBy, attributes, workspaceRec.isLocked, WorkspaceVersions.fromString(workspaceRec.workspaceVersion).getOrElse(throw new RawlsException(s"unexpected version string ${workspaceRec.workspaceVersion}")), workspaceRec.googleProject)
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:397]: make a `WorkspaceVersions.strictFromString` (or whatever you want to call it) that encapsulates the error message and throws if the version doesn't match? That way others can call it from elsewhere in the codebase and get the same error message.

**Review by davidangb — COMMENTED** [review]
lgtm - this isn't a full code review, more of an overview/high-level review

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 1 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:397]: I will get to it in a later PR

File: core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20201013_project_per_workspace.xml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    <changeSet logicalFilePath="dummy" author="dvoet" id="INVALID_BILLING_ACCT">
+        <addColumn tableName="BILLING_PROJECT">
+            <column name="INVALID_BILLING_ACCT" type="boolean" defaultValueBoolean="false">
+                <constraints nullable="false"/>
+            </column>
+        </addColumn>
+    </changeSet>
+    <changeSet logicalFilePath="dummy" author="dvoet" id="workspace_version">
+        <addColumn tableName="WORKSPACE">
+            <column name="WORKSPACE_VERSION" type="VARCHAR(10)" defaultValue="v1">
```
Comment (s-rubenstein) [inline core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20201013_project_per_workspace.xml:13]: What was the rationale behind making this a string rather than an int? (I could see int as a bit nicer because we could increment it more easily/have enforced formatting standards? Just curious rather than as a blocker.)

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala [hunk 3]
```diff
@@ -11,7 +11,7 @@ import org.broadinstitute.dsde.rawls.dataaccess.{GoogleBigQueryServiceFactory, M
 import org.broadinstitute.dsde.rawls.entities.EntityRequestArguments
 import org.broadinstitute.dsde.rawls.mock.{MockDataRepoDAO, MockSamDAO, MockWorkspaceManagerDAO}
 import org.broadinstitute.dsde.rawls.model.{DataReferenceName, RawlsUserEmail, UserInfo, Workspace}
-import org.joda.time.DateTime
+import org.joda.time. DateTime
```
Comment (s-rubenstein) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala]: Nit: This seems like an unrelated ` `?

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 4]
```diff
@@ -82,6 +84,25 @@ object AttributeName {
...
+  }
+  case object V2 extends WorkspaceVersion {
+    override val value: String = "v2"
+  }
+
+  def fromString(versionString: String): Option[WorkspaceVersion] = {
+    versionString match {
+      case V1.value => Option(V1)
+      case V2.value => Option(V2)
+      case _ => None
```
Comment (s-rubenstein) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:101]: Since this isn't nullable, should this error? (I don't know what code conventions are there yet)

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala [hunk 5]
```diff
@@ -644,6 +688,19 @@ class WorkspaceJsonSupport extends JsonSupport {
     }
   }
 
+  implicit object WorkspaceVersionFormat extends JsonFormat[WorkspaceVersion] {
+    override def write(wv: WorkspaceVersion): JsValue = JsString(wv.value)
+
+    override def read(json: JsValue): WorkspaceVersion = json match {
+      case JsString(versionString) => WorkspaceVersions.fromString(versionString) match {
+        case Some(version) => version
+        case None =>throw DeserializationException("unexpected version string")
```
Comment (s-rubenstein) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala]: Nit: I think there should be a ` ` after the `=>`? (Just given convention I've seen)

File: core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20201013_project_per_workspace.xml (hunk 2 above)
Comment (dvoet) [reply core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20201013_project_per_workspace.xml:13]: I have just envisioned it as v1 and v2 but it does not really matter. This is also not a long term thing. I don't anticipate a v3, etc.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala (hunk 3 above)
Comment (dvoet) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala]: fat fingers

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala (hunk 4 above)
Comment (dvoet) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:101]: this is just the string conversion code... it is up to the caller to determine if this is an error

**Review by s-rubenstein — APPROVED** [review]
I am happy with this, once the nits are addressed. Is there a ticket for the follow-on work that @davidangb mentioned? Or is that happening in the scope of this ticket?

**Review by marctalbott — APPROVED** [review]
